src/osd/OSD.cc

```diff
@@ -9474,14 +9474,15 @@ void OSDShard::consume_map(
         *pushes_to_free += qi.get_reserved_pushes();
 	slot->waiting.pop_front();
       }
-      if (slot->waiting.empty() &&
-	  slot->num_running == 0 &&
-	  !slot->pg) {
-	dout(20) << __func__ << "  " << pgid << " empty, pruning" << dendl;
-	p = pg_slots.erase(p);
-	continue;
-      }
     }
+    if (slot->waiting.empty() &&
```

Comment:
why? slot->waiting is updated in the if block above. so it is intuitive to check for empty in that if block also.